File: lib/feature/payment/infrastructure/persistence/order_repository_impl.dart
import 'dart:convert';
import 'package:dartz/dartz.dart';

import '../../../../shared/model/response_model.dart';
import '../../domain/model/response/orders.dart';
import '../../domain/repository/order_repository.dart';
import '../data_source/order_remote_data_source.dart';

class OrderRepositoryImpl implements OrderRepository {
  final OrderRemoteDataSource remoteDataSource;

  OrderRepositoryImpl({OrderRemoteDataSource? remoteDataSource})
      : remoteDataSource = remoteDataSource ?? OrderRemoteDataSource();

  @override
  Future<Either<ResponseModel, Orders>> getDetailOrder({
    required String reference,
  }) async {
    final response =
        await remoteDataSource.getDetailOrder(reference: reference);
    if (!response.isError) {
      return Right(
        Orders.fromJson(
          jsonDecode(response.result?.body ?? '')['data'],
        ),
      );
    }
    return Left(response);
  }

  @override
  Future<Either<ResponseModel, ResponseModel>> createOrderPayment({
    required String paymentMethod,
    required String reference,
  }) async {
    final response = await remoteDataSource.createOrderPayment(
      paymentMethod: paymentMethod,
      reference: reference,
    );
    if (!response.isError) {
      return Right(response);
    }
    return Left(response);
  }
}

File: src/features/payment/infrastructure/persistence/order_repository_impl.ts
import { Either, left, right } from "@/shared/utils/utility/either";
import { ResponseModel } from "@/shared/domain/model/response_model";
import { OrderRepository } from "../../domain/repository/order_repository";
import { OrderRemoteDataSource } from "../data_source/remote/order_remote_data_source";
import { OrdersResponse } from "../../domain/model/response/orders_response";
import { CreatePaymentRequest } from "../../domain/model/request/create_payment_request";

export class OrderRepositoryImpl implements OrderRepository {
  private dataSource: OrderRemoteDataSource;

  constructor(dataSource?: OrderRemoteDataSource) {
    this.dataSource = dataSource ?? new OrderRemoteDataSource();
  }

  async getDetailOrder(reference: string, token?: string): Promise<Either<ResponseModel, OrdersResponse>> {
    const res = await this.dataSource.getDetailOrder(reference, token);
    if (res.status && res.data) {
      return right(res.data);
    }
    return left(res);
  }

  async checkOrderStatus(reference: string, token?: string): Promise<Either<ResponseModel, any>> {
    const res = await this.dataSource.checkOrderStatus(reference, token);
    if (res.status) {
      return right(res.data);
    }
    return left(res);
  }

  async createPayment(request: CreatePaymentRequest, token?: string): Promise<Either<ResponseModel, any>> {
    const res = await this.dataSource.createPayment(request, token);
    if (res.status) {
      return right(res.data);
    }
    return left(res);
  }
}
